src/safeposix/syscalls/fs_calls.rs

```diff
@@ -2261,109 +2261,203 @@ impl Cage {
+    ///
+    /// ### Errors and Panics
+    ///
+    /// Currently, only one error is supported:
+    /// * `EINVAL` - the value of the mode argument is invalid.
+    /// Other errors, like `EFAULT  , `ENOTDIR`, etc. are not supported.
+    ///
+    /// There are no cases where this helper function panics.
+
+    pub fn _chmod_helper(inodenum: usize, mode: u32) -> i32 {
+        //S_IRWXA is a result of bitwise-or'ing read, write, and execute or search permissions for the file owner, group owners, 
+        //and other users. It encompasses all the mode bits that can be changed via `chmod_syscall()` and is used as a bitmask
+        //to make sure that no other invalid bit change is being made. 
+        if (mode & S_IRWXA) == mode {
+            //getting a mutable reference to an inode struct that corresponds to the file whose mode bits we want to change
+            let mut thisinode = FS_METADATA.inodetable.get_mut(&inodenum).unwrap();
+            //log is used to store all the changes made to the filesystem. After the cage is closed, all the collected
+            //changes are serialized and the state of the underlying filsystem is persisted. This allows us to avoid
+            //serializing and persisting filesystem state after every `chmod_syscall()`.
+            let mut log = true;
+            //We obtain the mode bits that should remain intact by bitwise-and'ing the inode's mode bits with 
+            //the set of bits that can be changed via `chmod_syscall`. The changes are applied by bitwise-or'ing 
+            //the intact mode bits with the changed mode bits.
             match *thisinode {
                 Inode::File(ref mut general_inode) => {
-                    general_inode.mode = (general_inode.mode & !S_IRWXA) | mode
+                    general_inode.mode = (general_inode.mode & !S_IRWXA) | mode;
                 }
                 Inode::CharDev(ref mut dev_inode) => {
                     dev_inode.mode = (dev_inode.mode & !S_IRWXA) | mode;
                 }
                 Inode::Socket(ref mut sock_inode) => {
                     sock_inode.mode = (sock_inode.mode & !S_IRWXA) | mode;
+                    //Sockets only exist as long as the cages using them are running. After these cages are closed, no changes
+                    //to sockets' inodes need to be persisted, thus using log is unnecessary.
                     log = false;
                 }
                 Inode::Dir(ref mut dir_inode) => {
                     dir_inode.mode = (dir_inode.mode & !S_IRWXA) | mode;
                 }
             }
+            //the mutable reference to the inode has to be dropped because log_metadata will need to acquire an immutable 
+            //reference to the same inode
             drop(thisinode);
+            //changes to an inode are saved into the log for all file types except for Sockets
             if log {
                 log_metadata(&FS_METADATA, inodenum)
```

Comment:
Ig `;` is missing at the end of this line.

Reply:
this brings no practical difference since neither log_metadata() nor _chmod_helper() return any values. nevertheless, added a ';' for completness